samza-api/src/main/java/org/apache/samza/metadatastore/MetadataStore.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.metadatastore;
+
+import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.container.SamzaContainerContext;
+
+import java.util.Map;
+
+/**
+ * Store abstraction responsible for managing the metadata of a Samza job and is agnostic of the
+ * deployment model (yarn/standalone) of the Samza job.
+ */
+@InterfaceStability.Evolving
+public interface MetadataStore {
+
+  /**
+   * Initializes the metadata store, if applicable, setting up the underlying resources
+   * and connections to the store endpoints. Upon successful completion of this method,
+   * metadata store is considered available to accept the client operations.
+   *
+   * @param containerContext represents the Samza container context.
+   */
+  void init(SamzaContainerContext containerContext);
```

Comment:
This can be used outside of SamzaContainer, and as such may not have access to the container context. Should provide access to Config and MetricsRegistry instead.

Reply:
+1 a CLI tool using this interface to read metadata wouldn't have this context.

<String, byte[]> or <String, Object> might be a simpler/better alternative for a starter impl.

It'd be nice if the metadatastore offered atomic updates across keys, e.g., for diagnostics a container-id and error-occurrence in that container should be recorded atomically. 

Does the implementation of this plan to internally use tableAPI implementation(s)?

Reply:
Makes sense. Removed Context and switched to having `config` and `metricsRegistry` as parameters. 

I'm not sure if it's a good idea to fix string as key type. This will prevent us in the future from persisting non-string keys in the metadatastore(since majority of store API's offer byte[], byte[] contract, this will incur a unnecessary limitation) and reading/writing it across samza-job, custom tools etc. 

>> It'd be nice if the metadatastore offered atomic updates across keys.

Not sure if it's a good idea to add this to interface contract and mandate this for every store implementation. For majority of scenarios in samza, a store implementation that persists the (key,value) through a synchronous remote call should be sufficient. 

Atomic updates through write-batch(or similar batch accumulation) is not supported by every store implementation. Unless we have prevalent and concrete use-cases across different layers in samza, i think it's a good idea not to introduce it.